src/expiring_lru.rs

```
@@ -0,0 +1,264 @@
+
+        this
+    }
+
+    pub fn insert(&self, key: K, value: V) {
+        match self.max_size {
+            Some(max_size) if max_size == 0 => return,
+            Some(max_size) if self.len() >= max_size => {
+                if let Some(lru) = self.get_lru() {
+                    let key = lru.key().clone();
+                    // We can't hold any references when removing something from the map
+                    drop(lru);
+                    self.remove(&key);
+
+                    debug!("Removed least recently used entry from expiring LRU");
+                }
+            }
+            _ => {}
+        }
+
+        let last_used = Instant::now();
+        let entry = Entry {
+            inner: value,
+            last_used,
+        };
+        self.inner.insert(key, entry);
+    }
+
+    pub fn get<Q>(&self, key: &Q) -> Option<EntryRef<'_, K, V>>
+    where
+        K: Borrow<Q>,
+        Q: Hash + Eq + ?Sized,
+    {
+        self.inner.get_mut(key).map(|mut entry| {
+            entry.last_used = Instant::now();
+
+            EntryRef(EntryRefInner::RefMut(entry))
+        })
+    }
+
+    pub fn get_lru(&self) -> Option<EntryRef<'_, K, V>> {
+        self.inner
+            .iter()
+            .next()
+            .map(|first_entry| {
+                self.inner.iter().fold(first_entry, |old, next| {
+                    if old.last_used > next.last_used {
+                        next
+                    } else {
+                        old
+                    }
+                })
+            })
+            .map(|entry| EntryRef(EntryRefInner::RefMulti(entry)))
+    }
+
+    pub fn remove(&self, key: &K) -> Option<(K, Entry<V>)> {
+        self.inner.remove(key)
+    }
+
+    pub fn len(&self) -> usize {
+        self.inner.len()
+    }
+
+    pub fn full(&self) -> bool {
+        match self.max_size {
+            Some(max_size) => self.len() >= max_size,
```

I'd rather have it panic should len >= max_size to catch bugs instead of silently continue "working"